src/main/java/com/amazon/opendistroforelasticsearch/security/support/WildcardMatcher.java

```diff
@@ -30,626 +30,503 @@
-     *
-     * @param pattern
-     * @param candidate
-     * @return
-     */
-    public static boolean matchAll(final String[] pattern, final String[] candidate) {
+        @Override
+        public boolean matchAny(Stream<String> candidates) {
+            return false;
+        }
 
+        @Override
+        public boolean matchAny(Collection<String> candidates) {
+            return false;
+        }
 
-        for (int i = 0; i < candidate.length; i++) {
-            final String string = candidate[i];
-            if (!matchAny(pattern, string)) {
-                return false;
-            }
+        @Override
+        public boolean matchAny(String[] candidates) {
+            return false;
         }
 
-        return true;
-    }
+        @Override
+        public boolean matchAll(Stream<String> candidates) {
+            return false;
+        }
 
-    /**
-     *
-     * @param pattern
-     * @param candidate
-     * @return
-     */
-    public static boolean allPatternsMatched(final String[] pattern, final String[] candidate) {
+        @Override
+        public boolean matchAll(Collection<String> candidates) {
+            return false;
+        }
 
-        int matchedPatternNum = 0;
+        @Override
+        public boolean matchAll(String[] candidates) {
+            return false;
+        }
 
-        for (int i = 0; i < pattern.length; i++) {
-            final String string = pattern[i];
-            if (matchAny(string, candidate)) {
-                matchedPatternNum++;
-            }
+        @Override
+        public <T extends Collection<String>> T getMatchAny(Stream<String> candidates, Collector<String, ?, T> collector) {
+            return Stream.<String>empty().collect(collector);
         }
 
-        return matchedPatternNum == pattern.length && pattern.length > 0;
-    }
+        @Override
+        public <T extends Collection<String>> T getMatchAny(Collection<String> candidate, Collector<String, ?, T> collector) {
+            return Stream.<String>empty().collect(collector);
+        }
 
-    public static boolean allPatternsMatched(final Collection<String> pattern, final Collection<String> candidate) {
+        @Override
+        public <T extends Collection<String>> T getMatchAny(String[] candidate, Collector<String, ?, T> collector) {
+            return Stream.<String>empty().collect(collector);
+        }
 
-        int matchedPatternNum = 0;
+        @Override
+        public boolean test(String candidate) {
+            return false;
+        }
 
-        for (String string:pattern) {
-            if (matchAny(string, candidate)) {
-                matchedPatternNum++;
-            }
+        @Override
+        public String toString() {
+            return "<NONE>";
         }
+    };
 
-        return matchedPatternNum == pattern.size() && pattern.size() > 0;
+    public static WildcardMatcher from(String pattern, boolean caseSensitive) {
```

Comment:
Is the main benefit of doing this to avoid recompilation of patterns which are expensive especially regexes ? 

But elasticsearch patterns use only wildcard match (specifically `*`) and other chars are restricted including `?` https://www.elastic.co/guide/en/kibana/current/index-patterns.html . Do we also want to restrict patterns similar to what elasticsearch supports as almost all the patterns we also have are wildcards? 

I see the SimpleMatcher which does actual wildcard matching still using pattern iteration for matching `*`. Wondering if I m missing any other benefits  ?

Reply:
Please check usage of `RegexMatcher`